lib/pg_ldap_sync/application.rb

```diff
@@ -150,21 +162,44 @@ def search_ldap_groups
         [name]
       end
 
+      group_members = if flatten_nested_groups
+        log.debug "Must flatten group"
+        retrieve_all_group_members(entry.dn)
+      else
+        retrieve_array_attribute(entry, member_attribute)
+      end
+
       names.each do |n|
-        group_members = retrieve_array_attribute(entry, member_attribute)
         groups << LdapRole.new(n, entry.dn, group_members)
       end
-      entry.each do |attribute, values|
-        log.debug "   #{attribute}:"
-        values.each do |value|
-          log.debug "      --->#{value.inspect}"
-        end
-      end
     end
     raise LdapError, "LDAP: #{@ldap.get_operation_result.message}" unless res
     return groups
   end
 
+  # Retrieves all users member of a group recursively using group's Distinguished Name
+  # It relies on LDAP_MATCHING_RULE_IN_CHAIN filter matching rule to flatten nested groups
+  #
+  # @param [String] group_dn group's distinguised name
+  # @return [array] array of users DN which are directly or indirectly member of the given group
+  def retrieve_all_group_members(group_dn)
+    ldap_user_conf = @config[:ldap_users]
+
+    log.debug "Retrieving members for #{group_dn}"
+
+    members = []
+    res = @ldap.search(
+      base: ldap_user_conf[:base],
+      filter: "(&(objectCategory=person)(objectClass=user)(memberOf:1.2.840.113556.1.4.1941:=#{group_dn}))",
```

Comment:
Can we remove these hardcoded values? I would like to keep only objectClass and memberOf.

Reply:
No problem.
Would you like to allow user to provide custom filter, or just make it simpler/minimal ?
I'm afraid it could become complex/error prone if filter is customisable.

Reply:
After a quick test, I've made the following change.
If `base` used for user request is a bit too "wide", I get computers when retrieving group members which seems odd.

```suggestion
      filter: "(&(!(objectClass=computer))(objectClass=user)(memberOf:1.2.840.113556.1.4.1941:=#{group_dn}))",
```